### PR TITLE
GraphQL: reset password with emailed token

### PR DESCRIPTION
**Pull request by sadakchap**
### New Pull Request Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am creating this PR in reference to an [issue](https://github.com/parse-community/parse-server/issues/7033).

### Issue Description

Related issue: resetting password with emailed token

### Approach
 - add one more mutation in [userMutation.js](https://github.com/parse-community/parse-server/blob/e88f2e38f9c38151869055878e52f1da8f7f6f99/src/GraphQL/loaders/usersMutations.js) file that will take username, password, token as arguments.
 - add one more route in [userRoutes.js](https://github.com/parse-community/parse-server/blob/e88f2e38f9c38151869055878e52f1da8f7f6f99/src/Routers/UsersRouter.js) file, that will add route POST `/resetPassword` that will basically call this [updatePassword](https://github.com/parse-community/parse-server/blob/e88f2e38f9c38151869055878e52f1da8f7f6f99/src/Controllers/UserController.js#L259) function

and with this, I think we should be able to complete reset password flow for GraphQL API also

### TODOs before merging

- [ ] Add test cases
- [ ] Add entry to changelog
- [ ] Add changes to documentation (guides, repository pages, in-code descriptions)
- [ ] Add [security check](https://github.com/parse-community/parse-server/blob/master/CONTRIBUTING.md#security-checks)
- [ ] Add new Parse Error codes to Parse JS SDK

### CODE REVIEW

File: src/Routers/UsersRouter.js [hunk 1]
```diff
@@ -357,6 +357,30 @@ export class UsersRouter extends ClassesRouter {
     );
   }
 
+  handleResetPassword(req) {
```
Comment (davimacedo) [inline src/Routers/UsersRouter.js:360]: It looks that the additions in this file are not necessary. You can use `resetPassword` function directly from the GraphQL API.

**Review by davimacedo — CHANGES_REQUESTED** [review]
Thanks for the PR!

Please find a comment below.

Also, could you please add a test case?

File: src/Routers/UsersRouter.js (hunk 1 above)
Comment (sadakchap) [reply src/Routers/UsersRouter.js:360]: Can you please point at which function you are talking about?? I am unable to find it

Comment (davimacedo) [reply src/Routers/UsersRouter.js:360]: This one: https://github.com/parse-community/parse-server/blob/master/src/Controllers/UserController.js#L260

**Comment by sadakchap** [comment]
@davimacedo about writing a test case, for faking email adaptor, I can see that in some test cases we are doing something like this
```js
const emailAdapter = {
  sendVerificationEmail: () => {},
  sendPasswordResetEmail: () => Promise.resolve(),
  sendMail: () => {},
};

// reconfiguring Server 
reconfigureServer({
  appName: 'test',
  emailAdapter: emailAdapter,
  publicServerURL: 'http://localhost:8378/1',
})
``` 
But still, how we will get the token, to actually test `resetPassword` mutation?

**Comment by davimacedo** [comment]
Instead of `sendPasswordResetEmail: () => Promise.resolve(),`, you can do:

`sendPasswordResetEmail: ({ appName, link, user }) => {
  // You can now, within this scope, retrieve the token from the link, test the new mutation, try to login with the new password, and finally call done() to finish the test
},`

File: src/Routers/UsersRouter.js (hunk 1 above)
Comment (sadakchap) [reply src/Routers/UsersRouter.js]: Hi @davimacedo , I updated both file userControllers.js & [userMutation.js](https://github.com/parse-community/parse-server/blob/e88f2e38f9c38151869055878e52f1da8f7f6f99/src/GraphQL/loaders/usersMutations.js), as suggested. Please let me know if more changes are needed.

**Comment by sadakchap** [comment]
> Instead of `sendPasswordResetEmail: () => Promise.resolve(),`, you can do:
> 
> `sendPasswordResetEmail: ({ appName, link, user }) => { // You can now, within this scope, retrieve the token from the link, test the new mutation, try to login with the new password, and finally call done() to finish the test },`

@davimacedo , I tried adding a test case as suggested. Please let me know if any changes are required.

**Review by davimacedo — APPROVED** [review]
It looks all good to me! @mtrezza this is a breaking change as we are changing the name os one of the current mutations. How should we proceed with that?

**Comment by mtrezza** [comment]
@davimacedo that depends on the reason for renaming `resetPasswordMutation` to `requestResetPasswordMutation`.

- If this is just cosmetic, I would ask whether the change is really necessary. A cosmetic change would probably not merit a breaking change. I would leave that to change organically in the future.

- If the change is justified, then preferably it should be phased-in, see [another example](https://github.com/parse-community/parse-server/issues/7292#issuecomment-803989620). If that is over the top for a single endpoint renaming, I would ask again whether the change is really necessary. If yes, the change would only be introduced with the next major Parse Server version. We do not have the new branch structures in place yet to hold-up a breaking change, so it may as well just be merged into master, but that is not how we would do it in the future.

**Comment by davimacedo** [comment]
It is actually kinda cosmetic. Let me explain. Currently we have a mutation called `resetPassword` that does not actually reset the password, but only sends the e-mail. This PR changes the name of this mutation to `requestResetPassword` and creates another mutation with the name `resetPassword` that receives the token (sent by e-mail) and the new password and actually resets the password. With this new mutation someone will be able to create a custom reset password form that could use the GraphQL API (and not the REST API) to complete flow. It will close the issue https://github.com/parse-community/parse-server/issues/7033.

But, maybe, to avoid the breaking change, we could keep the current mutation with the name `resetPassword` and create the new one with the name `confirmResetPassword` or something like this. The only advantage in changing the names is to be closer to what we currently have in the REST API and SDKs. What do you think?

**Comment by mtrezza** [comment]
To be pragmatic, am find with either way.

Actually, the current endpoints for password reset and email verification will be renamed with the new `PagesRouter`. The current endpoints are not semantically consistent. So aligning it probably isn't worth it for now.
 
My suggestion is to not make this a breaking change for now and rename all endpoints (and mutations) in one go with `6.0.0`. But I'm also fine with adding this breaking change to the master, if you rather do that.  

Up to release `5.0.0` I think we can be tolerant about breaking changes, because there are already some. From `5.0.0` onwards, we should follow the proposed versioning system and phased deprecation.

**Comment by davimacedo** [comment]
I agree. @sadakchap would you mind to change the mutations names to `resetPassword` and `confirmResetPassword`?

**Comment by sadakchap** [comment]
Right, @davimacedo changed the names of mutation both in userMutation file & test  also.

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [5.0.0-beta.1](https://github.com/parse-community/parse-server/releases/tag/5.0.0-beta.1)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [5.0.0](https://github.com/parse-community/parse-server/releases/tag/5.0.0)

**Pull request merged**